### PR TITLE
Bitcoins wallet implementation (for float64 data type)

### PR DESCRIPTION
**Pull request by Konstantsiy**

### CODE REVIEW

**Comment by Konstantsiy** [comment]
I packaged the tasks and made a module for the second task (the last 2 commits). Please tell me if there are any errors in the structure of the repository?

File: task2/go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module btcwallet
```
Comment (papahmsolo) [inline task2/go.mod]: By convention, module name should include the repo location
`github.com/Konstantsiy/go-internship/wallet`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Let's call the folders the same as modules 👍

File: task2/go.mod (hunk 1 above)
Comment (Konstantsiy) [reply task2/go.mod]: Ok, thanks for the correction.

File: btcwallet/btcwallet.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package btcwallet
```
Comment (papahmsolo) [inline btcwallet/btcwallet.go:2]: Let's add a package comment

File: btcwallet/btcwallet.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	balance    Bitcoin // balance stores the current value of the bitcoin(float64) balance in the wallet.
+}
+
+// NewBtcWallet returns a new BtcWallet which stores a predefined balance value.
+func NewBtcWallet(balance Bitcoin) *BtcWallet {
+	return &BtcWallet{balance: balance}
+}
+
+// Deposit increments the wallet balance by the given argument.
+func (wallet *BtcWallet) Deposit(data Bitcoin) error {
```
Comment (papahmsolo) [inline btcwallet/btcwallet.go]: Let's call the argument something like `b` or `amount`

File: btcwallet/btcwallet.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package btcwallet
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+)
```
Comment (papahmsolo) [inline btcwallet/btcwallet.go]: We can use `errors` package from standard lib instead

File: btcwallet/btcwallet_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package btcwallet
+
+import (
+	"sync"
+	"testing"
+
+	_ "golang.org/x/tools/go/expect"
+)
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: Not sure what is this for?

File: btcwallet/btcwallet_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+func TestBtcWallet(t *testing.T) {
+	wallet := NewBtcWallet(20)
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: Even though NewBtcWallet() accepts float as an argument, let's pass a value of type `Bitcoin`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool! Looks good!

Let's not worry about concurrency at the moment and implement good test for wallet first.
We'll need to test positive and negative scenarios for `Deposit`, `Withdraw`.

File: btcwallet/btcwallet.go (hunk 2 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet.go:2]: Sorry, a mistake from the last task

File: btcwallet/btcwallet.go (hunk 4 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet.go]: Okay, once again I trusted the IDE

File: btcwallet/btcwallet_test.go (hunk 5 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet_test.go]: Thanks, that's where so much garbage in go.sum comes from

File: btcwallet/btcwallet.go (hunk 3 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet.go]: Okay, I'll fix it now

File: btcwallet/btcwallet_test.go (hunk 6 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet_test.go]: Exactly, thank you

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

Let's add negative scenario test, such as expect error when trying to withdraw more than on the ballance

File: btcwallet/btcwallet_test.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+		if err := wallet.Withdraw(tc.Source); err != nil {
+			if err.Error() != tc.ExpectedErr.Error() {
+				t.Error(err.Error())
+			}
+		} else {
+			result := wallet.GetBalance().RoundTo(precision)
+			if result != tc.Expected {
+				t.Errorf("Incorrect result. Expect %.3f, got %.3f", tc.Expected, result)
+			}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: We can omit `else` in this case

File: btcwallet/btcwallet_test.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	const (
+		startBalance         Bitcoin = 1000
+		precision            int     = 3
+		withdrawErrorMessage string  = "the Bitcoins amount for the withdraw must be limited by the balance"
+	)
+
+	testTable := []struct {
+		Source      Bitcoin
+		Expected    Bitcoin
+		ExpectedErr error
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: Let's give `Source`  more meaningful name

File: btcwallet/btcwallet_test.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		},
+	}
+
+	for _, tc := range testTable {
+		wallet := NewBtcWallet(startBalance)
+
+		if err := wallet.Deposit(tc.Source); err != nil {
+			if err.Error() != tc.ExpectedErr.Error() {
+				t.Error(err.Error())
+			}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: We can create a helper function to assert error/no error, also mark the helper with `t.Helper()`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

We can add more info to error messages when tests fail on error, such as "expected error but didn't get one"

File: btcwallet/btcwallet_test.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+		go func(w *BtcWallet) {
+			defer wg.Done()
+			for j := 0; j < 2; j++ {
+				var err error
+				if err = w.Deposit(10); err != nil {
+					t.Errorf(err.Error())
+				}
+				if err = w.Withdraw(5); err != nil {
+					t.Errorf(err.Error())
+				}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: How about making deposits and withdrawals in the different goroutines?

File: btcwallet/btcwallet_test.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+func TestBtcWallet_Race(t *testing.T) {
+	var startBalance Bitcoin = 20
+	var expected Bitcoin = 50
+	wallet := NewBtcWallet(startBalance)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < 3; i++ {
+		wg.Add(1)
+		go func(w *BtcWallet) {
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: No need to pass the reference to the wallet, the goroutine will have access to it

File: btcwallet/btcwallet.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// Deposit increments the wallet balance by the given argument.
+func (wallet *BtcWallet) Deposit(amount Bitcoin) error {
+	wallet.Lock()
+	defer wallet.Unlock()
+
+	if amount <= 0 {
+		return errors.New("the Bitcoins amount for the deposit must be positive")
+	}
```
Comment (papahmsolo) [inline btcwallet/btcwallet.go]: We can create a package-level error and assert its value in the test instead of the error message. See `Errors and package APIs` section in this article https://blog.golang.org/go1.13-errors

File: btcwallet/btcwallet_test.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+		}
+	}
+}
+
+func errorAssertionHelper(t *testing.T, givenError, expectedError error) {
+	t.Helper()
+	if givenError.Error() != expectedError.Error() {
+		t.Error("expected error but didn't get one")
+	}
+}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: We can write the same helper for error, then we can `assertErr()` and `assertNoErr()` this will make our tests much more readable.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very nice!

Couple of issues to address:
- Try running `go test -race` to test data races
- Let's withdraw and deposit concurrently

File: btcwallet/btcwallet_test.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+func assertBalanceEquality(t *testing.T, result, expected Bitcoin) {
+	t.Helper()
+	if result != expected {
+		t.Errorf("Incorrect result. Expect %.3f, got %.3f", expected, result)
+	}
+}
+
+func TestBtcWallet_Race(t *testing.T) {
+	var startBalance Bitcoin = 20
+	var expected Bitcoin = 50
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: let's use chart variable declaration

Comment (Konstantsiy) [reply btcwallet/btcwallet_test.go]: > let's use chart variable declaration

Sorry, forgot ot put

File: btcwallet/btcwallet_test.go [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	"sync"
+	"testing"
+)
+
+func assertNoError(t *testing.T, err error) {
+	t.Helper()
+	if err != nil {
+		t.Errorf("No error expected but got: %v", err)
+	}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go:12]: This will always print `No error expected but got: nil` 😃

File: btcwallet/btcwallet_test.go [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+			defer wg.Done()
+			err := wallet.Deposit(10)
+			assertNoError(t, err)
+		}()
+		go func() {
+			defer wg.Done()
+			err := wallet.Withdraw(5)
+			assertNoError(t, err)
+		}()
+	}
```
Comment (papahmsolo) [inline btcwallet/btcwallet_test.go]: let's also call `wallet.GetBalance()` concurrently, it is causing the data race

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good! Final touches :)

File: btcwallet/btcwallet_test.go (hunk 15 above)
Comment (Konstantsiy) [reply btcwallet/btcwallet_test.go:12]: > This will always print `No error expected but got: nil`

*Stranger Things*

Comment (papahmsolo) [reply btcwallet/btcwallet_test.go:12]: My mistake

**Review by papahmsolo — COMMENTED** [review]
Nice work!

**Review by papahmsolo — APPROVED** [review]
